File: 572.subtree.of.another.tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} s
# @param {TreeNode} t
# @return {Boolean}
def is_subtree(s, t)
    def is_sub(s, t)
        if s==nil and t==nil then return true end
        if s==nil or t == nil or s.val != t.val then return false end
        if !is_sub(s.left, t.left) then return false end
        if !is_sub(s.right, t.right) then return false end
        return true
    end
    if is_sub(s, t) then return true end
    if s.left and is_subtree(s.left, t) then return true end
    if s.right and is_subtree(s.right, t) then return true end
    return false
end

File: 572. Subtree of Another Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:   
        if not root:
            return False
        
        def issub(n1, n2):
            if n1 == None and n2 == None:
                return True
            if (n1 == None) != (n2 == None):
                return False
            if n1.val != n2.val:
                return False
            return issub(n1.left, n2.left) and issub(n1.right, n2.right)
        
        return issub(root, subRoot) or self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
        
